tempesta_fw/http.c

```diff
@@ -941,12 +942,31 @@ tfw_http_req_zap_error(struct list_head *eq)
 
 	list_for_each_entry_safe(req, tmp, eq, fwd_list) {
 		list_del_init(&req->fwd_list);
-		tfw_http_error_resp_switch(req, req->httperr.status,
-					   req->httperr.reason);
+		if (!(req->flags & TFW_HTTP_F_REQ_DROP))
+			tfw_http_error_resp_switch(req, req->httperr.status,
+						   req->httperr.reason);
 		TFW_INC_STAT_BH(clnt.msgs_otherr);
 	}
 }
 
+/*
+ * If @req is dropped since the client fisconnected for some reason,
+ * move the request to the error queue @eq for sending an error response later.
+ */
+static inline bool
+tfw_http_req_evict_dropped(TfwSrvConn *srv_conn, TfwHttpReq *req,
+			   struct list_head *eq)
+{
+	if (unlikely(req->flags & TFW_HTTP_F_REQ_DROP)) {
+		TFW_DBG2("%s: Eviction: req=[%p] client disconnected\n",
+			 __func__, req);
+		tfw_http_req_err(srv_conn, req, eq, 403,
+				 "request evicted: client disconnected");
```

Comment:
Why do we need to pass the request to the error queue? `tfw_http_req_zap_error()` just does nothing on `req->flags & TFW_HTTP_F_REQ_DROP`, i.e. the `req` remains unfreed and we have a memory leakage. Moreover, we don't need to bother with the error queue - there is nothing to do with `TFW_HTTP_F_REQ_DROP` requests and it seems we can just free them right now.